### DISCUSSION TITLE
Save label contents to disk

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#138] [General]
I would like to save the contents of some labels to disk.

From what I understand the order of things is like this:

1: convert the ASCII to hex and put it in a table
2: merge tables, so create a table with all the individual tables in it
3: save to disk.

I have been reading various topics and looked at the Pro2 panel and I have not figured out a way 
to convert even 1 label content .

What is a beginners way of doing this?

Dnaldoog wrote this and it seems exactly what I need but how can I put that to work?

`str2hex(string): “string to hexa” = own function to convert a string to an hexa representation as a string
str2hex(“Hello”) gives “48656C6C6F”
In Ctrlr console you test with
console(tostring(str2hex("Hello")))
I’m using this function to store patch names in sysex.`

**Comment by dobo365** [comment]
Hi, this has been explained several times in the old forum so please also look there.
You may also need related functions so look in the Juce API for the File object.

In my Pro-2 panel there is a method called Save_onChange and another one SaveProgramData.

The way to analyze a panel is to start from a button, look at the OnChange method and then look further from there.
You should have found that as I'm also adding many comments...

Different steps are needed:
1. Create a file object that you will save into
2. Create and fill a memory block with data
3. Fill the file with the content of the memory block

```
f = utils.saveFileWindow ("Save text file", File(""), "*.txt", true)
if f:isValid() then
	if f:existsAsFile()==false then
		f:create()
	end	
	mbTemp = MemoryBlock(str2hex(modYourLabel:getComponent():getProperty("uiLabelText")))
	f:replaceWithData(mbTemp)
end
```

From there you can expand and handle anything ;-)
Good luck!

**Reply by dnaldoog** [reply]
[Save label text to binary_1_1_2021-01-21_18-02.zip](https://github.com/RomanKubiak/ctrlr/files/5848074/Save.label.text.to.binary_1_1_2021-01-21_18-02.zip)

Hi Baus,

Goodweather beat me to it!

The trick is to loop through each character of a string received from a label and then convert hex and save as binary, but how to do that?

See attached panel:

<pre style="color:steelblue">
function printBinaryToLabel(myText)
    -- Your method code here
    local word = {} -- initialize an empty table

    for i = 1, string.len(myText) do -- loop through each character in a string
        local w = myText:sub(i, i) -- extract the character at position i,i i.e. 2,2 returns the second character in myText
        word[i] = string.byte(w) -- write the byte representation to table (Could use table.insert() too)
    end
    <b>m</b> = MemoryBlock() -- create a block of Memory
    m:createFromTable(word) -- dump the table 'word' into the MemoryBlock
   end
</pre>

You then pass that MemoryBlock **m** object to be saved as file: Here I made it global, because the function is a callback to a uiButton method, so the MemoryBlock isn't passed as a parameter to the function, but I would usually do that to avoid global soup.

i.e.
<pre>
save=function(mBlock)

end
</pre>
<hr />
<pre>
save = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    if panel:getRestoreState() == true or panel:getProgramState() == true then
        return
    end
    if <b>m</b> ~= nil then
        file = utils.saveFileWindow("Save file", File.getSpecialLocation(File.userDesktopDirectory), "*.syx", true)
        if file:hasWriteAccess() then
            file:replaceWithData(<b>m</b>)
        end
    end
end

</pre>


... and there's probably an easier way of doing this! 

If you're adding multiple texts from different labels, I guess you would have to add them to one large string, but you would need some sort of delimiter (a character not in the allowed set of characters)  like '**|**'  <b style="color:blue"> = 7c</b> ( *ceci n'est pas une pipe*) for example, so then when you read the file back into your panel you know how to divide up and reassemble the binary string.

**Reply by dnaldoog** [reply]
Hi Baus,

Did you ever manage to get this sorted out, or maybe you didn't read this post?
